west: blobs: fetch `--allow-regex` filter

noahp opened the pull request:
When building in CI for specific SOCs, it's useful to only have `west blobs
fetch` pull a selected set of blob objects. This is especially helpful on
`hal_espressif`, which currently has 78 blob objects.

Add a `--allow-regex` arg to the `west blobs fetch` subcommand to filter
only specified blobs, for example:

```bash
# only download esp32 blobs, skip the other variants
❯ west blobs fetch hal_espressif --allow-regex 'lib/esp32/.*'
```

Also, replace all `str.format()` invocations with f-strings per review
feedback.

Signed-off-by: Noah Pendleton <noah.pendleton@gmail.com>

pdgendt commented on scripts/west_commands/blobs.py:
```diff
@@ -156,6 +161,12 @@ def fetch(self, args):
+
+            # if args.allow_regex is set, use it to filter the blob by path
+            if args.allow_regex and not re.match(args.allow_regex, blob['path']):
+                self.dbg('Blob {module}: {abspath} does not match regex '
+                         '{regex}, skipping'.format(regex=args.allow_regex, **blob))
```
Prefer using f-strings over `.format`.

noahp replied:
good call, done!

pdgendt commented on scripts/west_commands/blobs.py:
```diff
@@ -78,6 +79,13 @@ def do_add_parser(self, parser_adder):
+        group.add_argument(
+            '-l',
+            '--allow-regex',
+            help='''Regex pattern to apply to the blob local path.
+                            Only local paths matching this regex will be fetched.
+                            Note that local paths are relative to the module directory''',
```
Nit

```suggestion
            help='''Regex pattern to apply to the blob local path.
                    Only local paths matching this regex will be fetched.
                    Note that local paths are relative to the module directory''',
```

noahp commented:
@carlescufi / @swinslow any feedback on this change?

kartben reviewed (commented):
please add documentation to https://docs.zephyrproject.org/latest/develop/west/zephyr-cmds.html#working-with-binary-blobs-west-blobs as well

marc-hb commented on doc/develop/west/zephyr-cmds.rst:
```diff
@@ -214,6 +214,12 @@ Additionally the tool allows you to specify the modules you want to list,
+The argument ``--allow-regex`` can be passed ``west blobs fetch`` to restrict
+the specific blobs that are fetched, by passing a regular expression:
+
+  # only download esp32 blobs, skip the other variants
```
```suggestion
  # For example, only download esp32 blobs, skip the other variants
```

marc-hb reviewed (commented):
> Please do not combine formatting changes and functional changes in the same commit..

The OpenStack page is quite good on these topics:

https://wiki.openstack.org/wiki/GitCommitMessages#Things_to_avoid_when_creating_commits


Other good ones:

https://www.qemu.org/docs/master/devel/submitting-a-patch.html
https://www.kernel.org/doc/html/latest/process/submitting-patches.html

and of course:

https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html

noahp commented on doc/develop/west/zephyr-cmds.rst:
```diff
@@ -214,6 +214,12 @@ Additionally the tool allows you to specify the modules you want to list,
 parameter.
 
+The argument ``--allow-regex`` can be passed ``west blobs fetch`` to restrict
+the specific blobs that are fetched, by passing a regular expression:
```
:facepalm: ohh thank you! I'm blind, can't believe I missed it!

noahp committed:
west: blobs: fetch `--allow-regex` filter

When building in CI for specific SOCs, it's useful to only have `west blobs
fetch` pull a selected set of blob objects. This is especially helpful on
`hal_espressif`, which currently has 78 blob objects.

Add a `--allow-regex` arg to the `west blobs fetch` subcommand to filter
only specified blobs, for example:

```bash
# only download esp32 blobs, skip the other variants
❯ west blobs fetch hal_espressif --allow-regex 'lib/esp32/.*'
```

Also, replace all `str.format()` invocations with f-strings per review
feedback.

Signed-off-by: Noah Pendleton <noah.pendleton@gmail.com>
